tree/ntuple/v7/src/RNTupleJoinTable.cxx

```diff
@@ -78,51 +61,140 @@ void ROOT::Experimental::Internal::RNTupleJoinTable::Build(RPageSource &pageSour
    }
 
-   auto entryIdxs = fJoinTable.find(RCombinedJoinFieldValue(joinFieldValues));
+   if (const auto &entries = fMapping.find(RCombinedJoinFieldValue(joinFieldValues)); entries != fMapping.end()) {
+      return &entries->second;
+   }
+
+   return nullptr;
+}
+
+//------------------------------------------------------------------------------
+
+std::unique_ptr<ROOT::Experimental::Internal::RNTupleJoinTable>
+ROOT::Experimental::Internal::RNTupleJoinTable::Create(const std::vector<std::string> &fieldNames)
+{
+   return std::unique_ptr<RNTupleJoinTable>(new RNTupleJoinTable(fieldNames));
+}
+
+std::vector<ROOT::Experimental::Internal::RNTupleJoinTable::REntryMapping *>
+ROOT::Experimental::Internal::RNTupleJoinTable::GetAllMappings() const
+{
+   std::vector<REntryMapping *> joinMappingPtrs;
+
+   for (const auto &key : fPartitionKeys) {
+      auto mappings = GetMappingsForKey(key);
+      joinMappingPtrs.insert(joinMappingPtrs.end(), mappings.begin(), mappings.end());
+   }
+
+   return joinMappingPtrs;
+}
 
-   if (entryIdxs == fJoinTable.end())
+std::vector<ROOT::Experimental::Internal::RNTupleJoinTable::REntryMapping *>
+ROOT::Experimental::Internal::RNTupleJoinTable::GetMappingsForKey(PartitionKey_t partitionKey) const
+{
+   auto partitionIter = std::find(fPartitionKeys.begin(), fPartitionKeys.end(), partitionKey);
+   if (partitionIter == fPartitionKeys.end())
       return {};
 
-   return entryIdxs->second;
+   auto partitionIdx = std::distance(fPartitionKeys.begin(), partitionIter);
+   std::vector<REntryMapping *> joinMappingPtrs;
+   joinMappingPtrs.reserve(fPartitions[partitionIdx].size());
+
+   for (const auto &joinMapping : fPartitions[partitionIdx]) {
+      joinMappingPtrs.emplace_back(joinMapping.get());
+   }
+   return joinMappingPtrs;
+}
+
+ROOT::Experimental::Internal::RNTupleJoinTable &
+ROOT::Experimental::Internal::RNTupleJoinTable::Add(RPageSource &pageSource, PartitionKey_t partitionKey)
+{
+   if (fIsBuilt) {
+      throw RException(R__FAIL("cannot add to an already-built join table"));
+   }
+
+   auto joinMapping = std::unique_ptr<REntryMapping>(new REntryMapping(pageSource, fJoinFieldNames));
+
+   auto partitionIter = std::find(fPartitionKeys.begin(), fPartitionKeys.end(), partitionKey);
+   if (partitionIter == fPartitionKeys.end()) {
+      std::vector<std::unique_ptr<REntryMapping>> newPartition;
+      newPartition.emplace_back(std::move(joinMapping));
+      fPartitions.emplace_back(std::move(newPartition));
+      fPartitionKeys.push_back(partitionKey);
+   } else {
+      auto partitionIdx = std::distance(fPartitionKeys.begin(), partitionIter);
+      fPartitions[partitionIdx].emplace_back(std::move(joinMapping));
+   }
+
+   return *this;
+}
+
+void ROOT::Experimental::Internal::RNTupleJoinTable::Build()
+{
+   if (fIsBuilt)
+      return;
+
+   for (auto &mapping : GetAllMappings()) {
+      mapping->Build();
+   }
```

Comment:
Given that `GetAllMappings` creates a copy of the pointers stored in `fPartitions`, should this just iterate over it directly?
```suggestion
   for (const auto &partition : fPartitions) {
      for (const auto &joinMapping : partition) {
         joinMapping->Build();
      }
   }
```